### DISCUSSION TITLE
about unetr

### DISCUSSION THREAD

**Discussion by jhkang0526** [Project-MONAI/MONAI#2547] [Q&A] [answered]
hello,

I recently saw about the UNETR. 

Is it currently possible to apply to 3D data?

If possible, how should I use it?

Thanks,

**Comment by Nic-Ma** [comment]
Hi @ahatamiz ,

Could you please help share some comments here?

Thanks in advance.

**Reply by ahatamiz** [reply]
Hi @jhkang0526 , @Nic-Ma 

Yes, it directly uses 3D data. I am preparing tutorial for UNETR. Hopefully it can be useful. 

Thanks

**Reply by jhkang0526** [reply]
Let me ask you a dumb question. 
Error occurs when import UNETR using this(from monai.networks.nets import UNETR). How can this be solved? 

It is said that it is not included in the __init__.
`ImportError: cannot import name 'UNETR' from 'monai.networks.nets' (/home/kjh/anaconda3/envs/kjh/lib/python3.8/site-packages/monai/networks/nets/__init__.py)
`

**Reply by ahatamiz** [reply]
UNETR was recently added. Would you please install the latest build and let me know if there is still an issue ? 

I am currently able to import it with the latest version (MONAI version: 0.6.dev2127).

**Reply by jhkang0526** [reply]
oh, sorry! this is my mistake.

I thought this was the latest version(0.5.3).
 I'm not sure if 0.6.dev2127 is the same as 0.6.0rc2, but when I reinstall it(0.6.0rc2), I can import that.
(from monai.networks.nets.unetr import UNETR)

Is this the right way to proceed?

Thanks,

**Reply by ahatamiz** [reply]
Great. Glad it can be imported properly. I am going to add the doc strings as well as the tutorials, but until it's added, this is an example of a UNETR for multi-organ segmentation (14 classes, one input channel CT image) with input patch size of 96:

```
model = UNETR(
    in_channels=1, 
    out_channels=14, 
    img_size=(96,96,96), 
    feature_size=16,
    hidden_size=768,
    mlp_dim=3072, 
    num_heads=12, 
    pos_embed='perceptron', 
    norm_name='instance',
    res_block=True, 
    dropout_rate=0.0).to(device)
```

Full tutorial will be provided.

**Reply by jhkang0526** [reply]
Thank you very much! 
I hope the day will come when I can share the results with you using this!

**Reply by ahatamiz** [reply]
Thank you @jhkang0526 .

**Reply by dianemarquette** [reply]
Hi @ahatamiz ! Thanks for the implementation of UNETR. I look forward to trying it out on my dataset :). Do you have an approximate release date for the tutorial?

**Reply by ahatamiz** [reply]
Hi @dianemarquette  thank you for your interest in UNETR. We should be able to release the UNETR tutorial next week. The tutorial is currently made for mulit-organ segmentation using BTCV dataset. We plan to add tutorials for other segmentation tasks in the future. 

Thanks

**Comment by wyli** [comment]
I'm converting this into a feature request for docstring enhancement. Would be great to have examples in the docstring such as:
https://github.com/Project-MONAI/MONAI/blob/b4fca1d12245e8d7950ae4782058dc3cea6a9775/monai/networks/nets/basic_unet.py#L174-L188

**Reply by ahatamiz** [reply]
Hi @wyli Sounds like a great idea. I can add.

**Comment by ahatamiz** [comment, marked as answer]
Hi @dianemarquette  and @jhkang0526

A tutorial for UNETR on 3D multi-organ segmentation using BTCV dataset is now provided in MONAI tutorials:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/unetr_btcv_segmentation_3d.ipynb

Additionally, UNETR repository is now available in the following:
https://monai.io/research/unetr-btcv-multi-organ-segmentation

Thanks

**Reply by jhkang0526** [reply]
Hi @ahatamiz 

Thank you for uploading the tutorial so quickly!

**Reply by dianemarquette** [reply]
Thank you @ahatamiz !

**Comment by LalehSeyyed** [comment]
@ahatamiz I am wondering in 3D images if the size of the images in z-axis varies (which is very common) what number we should use for img_size. For instance, my 3D image sizes are (512,512, z) where z may vary in a large range e.g. 40 to 500. 

 Additionally, for the other setting, where can we find the logic behind how to select them if there is a dependency between the numbers?

**Reply by ahatamiz** [reply]
Hi @LalehSeyyed, currently UNETR supports fixed sized inputs. In order to handle images with various sizes, we recommend using a sampler with fixed size or resizing to a predefined resolution. For instance in the BTCV tutorial, we use a fixed size sampler. 
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/unetr_btcv_segmentation_3d.ipynb

The img_size will then be the same as the predefined size that is used for sampling the data (e.g. (96,96,96) for BTCV). It is a hyper-parameter which needs tuning for each specific task. For brain tumor segmentation (BRATS) for instance, we set it to (128,128,128) to better capture tumors of variable sizes.

**Comment by LalehSeyyed** [comment]
Hi @ahatamiz
Thanks for your response. I am wondering if there is any pre-trained model available based on UNETR. Our datasets are small and starting from a retained model may help the network to converge. Currently, I can not learn due to the size issue.

Best
Laleh

**Comment by ahatamiz** [comment]
Hi @LalehSeyyed 

We provide pre-trained weights along with detailed fine-tuning instructions in the UNETR MONAI Research Contribution Repository:
https://github.com/Project-MONAI/research-contributions/tree/master/UNETR/BTCV

Thanks,
Ali
